packages/flutter/lib/src/rendering/paragraph.dart

```diff
@@ -644,7 +650,7 @@ class RenderParagraph extends RenderBox
     size = constraints.constrain(textSize);
 
     final bool didOverflowHeight = size.height < textSize.height || textDidExceedMaxLines;
-    final bool didOverflowWidth = size.width < textSize.width;
+    final bool didOverflowWidth = size.width < textSize.width || size.width < _textPainter.longestLine;
```

Comment:
Why would  `_textPainter.longestLine` ever be smaller than `textSize.width` (which is `_textPainter.size,width`) for a one char long text? That seems odd. Is there a bug in how textPainter calculates its actual width?

Reply:
We also have `TextWidthBasis` which essentially changed `_textPainter.size.width` to be `_textPainter.longestLine`. I am not too familiar with that setting, though.

/cc @justinmc who added it a while back.

Reply:
> Why would `_textPainter.longestLine` ever be smaller than `textSize.width` (which is `_textPainter.size,width`) for a one char long text? That seems odd. Is there a bug in how textPainter calculates its actual width?

The `textPainter.size.width` always returns the layout `constraints.width`, this should be the implementation of the engine layer. @LongCatIsLooong  may know the details.

Reply:
_textPainter.longestLine is the width of the longest line of actual text.  _textPainter.width is either the same thing or the constraints width depending on TextWidthBasis ([code](https://github.com/flutter/flutter/blob/b7a18be0a027805705689827376c578d98d53012/packages/flutter/lib/src/painting/text_painter.dart#L538)) as you guys noted.

So it's possible and even common for _textPainter.longestLine to be shorter than _textPainter.width, for example:

<img width="138" alt="Screen Shot 2022-03-22 at 9 23 40 AM" src="https://user-images.githubusercontent.com/389558/159526748-a501c56a-ff55-4097-b751-3d0cfb52ec1c.png">

Here _textPainter.longestLine is just the width of that single character (a few pixels), while _textPainter.width is the width of the green Container (100px).

So the condition in the code here (`size.width < _textPainter.longestLine`) makes sense to me.  That will be true when the available space is less wide than a single character.

Reply:
PS Sorry it took me so long to respond here!  I'm bad with my notifications recently.